cpp/DataStorm/sampleFilter/Writer.cpp

```diff
@@ -3,44 +3,85 @@
 #include <DataStorm/DataStorm.h>
 #include <Ice/Ice.h>
 
+#include "TemperatureRange.h"
+
 #include <iostream>
-#include <optional>
 #include <string>
 
 using namespace std;
 
 int
 main(int argc, char* argv[])
 {
-    try
-    {
-        // Instantiates DataStorm node.
-        DataStorm::Node node{argc, argv};
+    // CtrlCHandler must be created before the communicator is created or any other threads are started.
+    Ice::CtrlCHandler ctrlCHandler;
 
-        // Instantiates the "hello" topic. The topic uses strings for keys and values.
-        DataStorm::Topic<string, string> topic{node, "hello"};
+    // Instantiates DataStorm node.
+    DataStorm::Node node{argc, argv};
 
-        // Configure writers to not clear the history. We want the readers to receive all the writer samples.
-        topic.setWriterDefaultConfig({std::nullopt, std::nullopt, DataStorm::ClearHistoryPolicy::Never});
+    // Instantiates the "temperature" topic. The topic uses strings for keys and floats for values.
+    DataStorm::Topic<string, float> topic{node, "temperature"};
 
-        // Instantiate the foo writer and wait for a reader to connect.
-        auto writer = DataStorm::makeSingleKeyWriter(topic, "foo");
-        topic.waitForReaders();
+    // Create a sample filter that only reports temperatures outside of the given temperature range.
+    // The filter criteria TemperatureRange is defined in TemperatureRange.ice Slice file.
+    topic.setSampleFilter<ClearSky::TemperatureRange>(
+        "not-in-range-filter",
+        [](ClearSky::TemperatureRange range)
+        {
+            return [range](const DataStorm::Sample<string, float>& sample)
+            {
+                float value = sample.getValue();
+                return value < range.min || value > range.max;
+            };
+        });
 
-        // Publish samples
-        writer.update("hi");
-        writer.update("greetings");
-        writer.update("good morning");
-        writer.update("hello");
-        writer.update("good afternoon");
+    // Instantiate an any key writer.
+    auto writer = DataStorm::makeAnyKeyWriter(topic, "temperature-writer");
 
-        // Wait for readers to disconnect.
-        topic.waitForNoReaders();
-    }
-    catch (const std::exception& ex)
+    // Wait for a reader to connect
+    topic.waitForReaders();
+
+    // Shutdown the writer on Ctrl-C.
+    auto shutdownPromise = std::promise<void>();
+    auto shutdownFuture = shutdownPromise.get_future();
+    ctrlCHandler.setCallback(
+        [&ctrlCHandler, &shutdownPromise](int)
+        {
+            std::cout << "Shutting down..." << std::endl;
+            shutdownPromise.set_value();
+            // Reset the callback to nullptr to avoid calling it again.
+            ctrlCHandler.setCallback(nullptr);
```

Comment:
I'm also dubious of the `setCallback` mutex locking here.

Here, it makes more sense, since `set_value` isn't allowed to be called multiple times.
But if the user is spamming Ctrl-C over and over, I think it's fine to die with an exception.

With the current logic, it sure looks to me like spamming Ctrl-C could cause a dead-lock in `setCallback`
(if a 2nd interrupt is triggered while the handler is still responding to a 1st signal).